google/cloud/connection_options.h

```diff
@@ -0,0 +1,243 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_CONNECTION_OPTIONS_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_CONNECTION_OPTIONS_H
+
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/internal/background_threads_impl.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/tracing_options.h"
+#include <grpcpp/grpcpp.h>
+#include <functional>
+#include <memory>
+#include <set>
+#include <string>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+std::set<std::string> DefaultTracingComponents();
+TracingOptions DefaultTracingOptions();
+void DefaultLogging();
+std::unique_ptr<BackgroundThreads> DefaultBackgroundThreads();
+}  // namespace internal
+
+/**
+ * The configuration parameters for spanner connections.
+ */
+template <typename ConnectionTraits>
+class ConnectionOptions {
+ public:
+  /// The default options, using `grpc::GoogleDefaultCredentials()`.
+  ConnectionOptions() : ConnectionOptions(grpc::GoogleDefaultCredentials()) {}
+
+  /// Default parameters, using an explicit credentials object.
+  explicit ConnectionOptions(
+      std::shared_ptr<grpc::ChannelCredentials> credentials)
+      : credentials_(std::move(credentials)),
+        endpoint_(ConnectionTraits::default_endpoint()),
+        num_channels_(ConnectionTraits::default_num_channels()),
+        tracing_components_(internal::DefaultTracingComponents()),
+        tracing_options_(internal::DefaultTracingOptions()),
+        user_agent_prefix_(ConnectionTraits::user_agent_prefix()),
+        background_threads_factory_(internal::DefaultBackgroundThreads) {
+    internal::DefaultLogging();
+  }
+
+  /// Change the gRPC credentials from the default of
```

Comment:
Re "from the default of ...": `credentials_` need not have the default value here.

Reply:
Fixed.